### ISSUE TITLE
On linux with a touch screen, GestureDetector detects vertical drag instead of a horizontal drag.

### ISSUE THREAD

**Issue by jdebou** [flutter/flutter#78232]
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update main.dart as follows: 
<details>
  <summary>main.dart</summary> 

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _hCounter = 0;
  int _vCounter = 0;
 
  void _incrementVCounter() {
    print("vertical");
    setState(() {
      _vCounter++;
    });
  }
  void _incrementHCounter() {
    print("horizontal");
    setState(() {
      _hCounter++;
    });
  }

  // Build a GestureDetector on the whole scaffold body and listen for horizontal
  // and vertical drags.
  // Display the number of drags for each type.
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        width: double.infinity,
        height: double.infinity,
        child: GestureDetector(
          behavior: HitTestBehavior.translucent,
          onHorizontalDragStart: (_) => _incrementHCounter(),
          onVerticalDragStart: (_) => _incrementVCounter(),
          child: Center(
            child: Column(
              children: [
                Container(height: 64),
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Text(
                      'You have triggered vertical drag this many times:',
                    ),
                    Text(
                      '$_vCounter',
                      style: Theme.of(context).textTheme.headline4,
                    ),
                  ],
                ),
                Container(height: 64),
                Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Text(
                      'You have triggered horizontal drag this many times:',
                    ),
                    Text(
                      '$_hCounter',
                      style: Theme.of(context).textTheme.headline4,
                    ),
                  ],
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

3. Swipe horizontally

**Expected results:** 

I expected the `onHorizontalDragStart` callback to be called each time I start a horizontal drag.

**Actual results:** 

90% of the time, the `onVerticalDragStart` callback was called instead.

<details>
  <summary>Logs</summary>

```
 bug flutter run --verbose
[  +51 ms] executing: [/home/joackim/snap/flutter/common/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +23 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] c5a4b4029c0798f37c4a39b479d7cb75daa7b05c
[        ] executing: [/home/joackim/snap/flutter/common/flutter/] git tag --points-at c5a4b4029c0798f37c4a39b479d7cb75daa7b05c
[   +9 ms] Exit code 0 from: git tag --points-at c5a4b4029c0798f37c4a39b479d7cb75daa7b05c
[        ] 2.0.1
[  +27 ms] executing: [/home/joackim/snap/flutter/common/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/beta
[        ] executing: [/home/joackim/snap/flutter/common/flutter/] git ls-remote --get-url origin
[   +4 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +30 ms] executing: [/home/joackim/snap/flutter/common/flutter/] git rev-parse --abbrev-ref HEAD
[   +5 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] beta
[  +34 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +42 ms] executing: /home/joackim/Android/Sdk/platform-tools/adb devices -l
[  +34 ms] List of devices attached
[   +2 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +56 ms] Skipping pub get: version match.
[  +86 ms] Generating /home/joackim/dev/bug/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[  +61 ms] Initializing file store
[   +7 ms] Skipping target: gen_localizations
[   +4 ms] complete
[   +3 ms] Launching lib/main.dart on Linux in debug mode...
[   +3 ms] /home/joackim/snap/flutter/common/flutter/bin/cache/dart-sdk/bin/dart --disable-dart-dev /home/joackim/snap/flutter/common/flutter/bin/cache/artifacts/engine/linux-x64/frontend_server.dart.snapshot --sdk-root /home/joackim/snap/flutter/common/flutter/bin/cache/artifacts/engine/common/flutter_patched_sdk/ --incremental --target=flutter
--debugger-module-names --experimental-emit-debug-metadata --output-dill /tmp/flutter_tools.BZQNFR/flutter_tool.RVXLRC/app.dill --packages /home/joackim/dev/bug/.dart_tool/package_config.json -Ddart.vm.profile=false -Ddart.vm.product=false --enable-asserts --track-widget-creation --filesystem-scheme org-dartlang-root --initialize-from-dill
build/cache.dill.track.dill
[  +15 ms] Building Linux application...
[  +14 ms] <- compile package:bug/main.dart
[   +2 ms] executing: [build/linux/debug/] cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug /home/joackim/dev/bug/linux
[  +38 ms] -- Configuring done
[   +3 ms] -- Generating done
[        ] -- Build files have been written to: /home/joackim/dev/bug/build/linux/debug
[   +3 ms] executing: ninja -C build/linux/debug install
[  +10 ms] ninja: Entering directory `build/linux/debug'
[+2834 ms] [1/5] Generating /home/joackim/dev/bug/linux/flutter/ephemeral/libflutter_linux_gtk.so, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_basic_message_channel.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_binary_codec.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_binary_messenger.h,
/home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_dart_project.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_engine.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_json_message_codec.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_json_method_codec.h,
/home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_message_codec.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_method_call.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_method_channel.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_method_codec.h,
/home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_method_response.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_plugin_registrar.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_plugin_registry.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_standard_message_codec.h,
/home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_standard_method_codec.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_string_codec.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_value.h, /home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/fl_view.h,
/home/joackim/dev/bug/linux/flutter/ephemeral/flutter_linux/flutter_linux.h, _phony_
[   +2 ms] [  +55 ms] executing: [/home/joackim/snap/flutter/common/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] [  +25 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[        ] [        ] c5a4b4029c0798f37c4a39b479d7cb75daa7b05c
[        ] [        ] executing: [/home/joackim/snap/flutter/common/flutter/] git tag --points-at c5a4b4029c0798f37c4a39b479d7cb75daa7b05c
[        ] [   +8 ms] Exit code 0 from: git tag --points-at c5a4b4029c0798f37c4a39b479d7cb75daa7b05c
[        ] [        ] 2.0.1
[        ] [  +27 ms] executing: [/home/joackim/snap/flutter/common/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[        ] [   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] [        ] origin/beta
[        ] [        ] executing: [/home/joackim/snap/flutter/common/flutter/] git ls-remote --get-url origin
[        ] [   +2 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] [        ] https://github.com/flutter/flutter.git
[        ] [  +26 ms] executing: [/home/joackim/snap/flutter/common/flutter/] git rev-parse --abbrev-ref HEAD
[        ] [   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] [        ] beta
[        ] [  +34 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ] [   +2 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[        ] [  +41 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'GradleWrapper' is not required, skipping update.
[        ] [        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterSdk' is not required, skipping update.
[        ] [        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping update.
[        ] [        ] Artifact Instance of 'PubDependencies' is not required, skipping update.
[        ] [  +75 ms] Initializing file store
[        ] [   +5 ms] Done initializing file store
[        ] [  +27 ms] Skipping target: gen_localizations
[        ] [+1305 ms] Skipping target: unpack_linux
[        ] [ +335 ms] Skipping target: kernel_snapshot
[        ] [ +211 ms] Skipping target: debug_bundle_linux_assets
[        ] [        ] Persisting file store
[        ] [   +5 ms] Done persisting file store
[        ] [   +3 ms] build succeeded.
[        ] [   +6 ms] "flutter assemble" took 2,028ms.
[        ] [ +185 ms] ensureAnalyticsSent: 182ms
[        ] [   +2 ms] Running shutdown hooks
[        ] [        ] Shutdown hooks complete
[        ] [        ] exiting with code 0
[ +270 ms] [2/5] Building CXX object CMakeFiles/bug.dir/flutter/generated_plugin_registrant.cc.o
[  +30 ms] [3/5] Building CXX object CMakeFiles/bug.dir/my_application.cc.o
[  +68 ms] [4/5] Linking CXX executable intermediates_do_not_run/bug
[        ] [4/5] Install the project...
[   +8 ms] -- Install configuration: "Debug"
[  +17 ms] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/bug
[        ] -- Set runtime path of "/home/joackim/dev/bug/build/linux/debug/bundle/bug" to "$ORIGIN/lib"
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/icudtl.dat
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/lib/libflutter_linux_gtk.so
[  +93 ms] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/version.json
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/packages
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/packages/cupertino_icons
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/packages/cupertino_icons/assets
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/packages/cupertino_icons/assets/CupertinoIcons.ttf
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/fonts
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/fonts/MaterialIcons-Regular.otf
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/AssetManifest.json
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/NOTICES
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/FontManifest.json
[        ] -- Installing: /home/joackim/dev/bug/build/linux/debug/bundle/data/flutter_assets/kernel_blob.bin
[  +39 ms] Building Linux application... (completed in 3.4s)
[ +189 ms] Observatory URL on device: http://127.0.0.1:36029/R7WcboEaYq4=/
[   +3 ms] Caching compiled dill
[  +39 ms] Connecting to service protocol: http://127.0.0.1:36029/R7WcboEaYq4=/
[ +106 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0, connecting to VM service at http://127.0.0.1:36029/R7WcboEaYq4=/.
[  +67 ms] DDS is listening at http://127.0.0.1:39985/naykAwWupcI=/.
[  +56 ms] Successfully connected to service protocol: http://127.0.0.1:36029/R7WcboEaYq4=/
[ +679 ms] DevFS: Creating new filesystem on the device (null)
[  +10 ms] DevFS: Created new filesystem on the device (file:///tmp/bugVFHLRZ/bug/)
[   +3 ms] Updating assets
[  +62 ms] Syncing files to device Linux...
[   +1 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +1 ms] <- recompile package:bug/main.dart d3540f8d-c322-4c77-8d96-3001871c2775
[        ] <- d3540f8d-c322-4c77-8d96-3001871c2775
[  +73 ms] Updating files.
[        ] DevFS: Sync finished
[   +1 ms] Syncing files to device Linux... (completed in 77ms)
[        ] Synced 0.0MB.
[        ] <- accept
[   +4 ms] Connected to _flutterView/0x27ac570.
[   +1 ms] Flutter run key commands.
[   +1 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] An Observatory debugger and profiler on Linux is available at: http://127.0.0.1:39985/naykAwWupcI=/
[        ] 
           Flutter DevTools, a Flutter debugger and profiler, on Linux is available at: http://127.0.0.1:9101?uri=http%3A%2F%2F127.0.0.1%3A39985%2FnaykAwWupcI%3D%2F
[        ] Running with unsound null safety
[        ] For more information see https://dart.dev/null-safety/unsound-null-safety
[+2832 ms] flutter: vertical
[ +336 ms] flutter: vertical
[ +383 ms] flutter: vertical
[ +383 ms] flutter: vertical
[ +400 ms] flutter: vertical
[ +455 ms] flutter: vertical
[ +412 ms] flutter: vertical
[ +449 ms] flutter: vertical
[ +366 ms] flutter: vertical
[ +416 ms] flutter: vertical
[ +371 ms] flutter: vertical
[ +379 ms] flutter: vertical
[ +384 ms] flutter: vertical
[ +331 ms] flutter: vertical
[ +403 ms] flutter: vertical
[ +431 ms] flutter: vertical
[ +382 ms] flutter: vertical
[ +434 ms] flutter: vertical
[ +433 ms] flutter: vertical
[ +333 ms] flutter: vertical
[+5134 ms] Application finished.

```

```
bug flutter analyze
Analyzing bug...                                                        
No issues found! (ran in 1.1s)
```

```
flutter doctor -v
[✓] Flutter (Channel beta, 2.0.1, on Linux, locale en_US.UTF-8)
    • Flutter version 2.0.1 at /home/joackim/snap/flutter/common/flutter
    • Framework revision c5a4b4029c (11 days ago), 2021-03-04 09:47:48 -0800
    • Engine revision 40441def69
    • Dart version 2.12.0

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /home/joackim/Android/Sdk
    • Platform android-30, build-tools 30.0.2
    ✗ Android SDK file not found: /home/joackim/Android/Sdk/platforms/android-30/android.jar.
    • Try re-installing or updating your Android SDK,
      visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions.

[✗] Chrome - develop for the web (Cannot find Chrome executable at /home/joackim/chrome-unsafe/google-chrome-unsafe.sh)
    ! /home/joackim/chrome-unsafe/google-chrome-unsafe.sh is not executable.

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/linux#android-setup for detailed instructions).

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-x64 • Linux

! Doctor found issues in 3 categories.

```

</details>

**Comment by darshankawar** [comment]
Tagging @markusaksli-nc to verify this on Linux.

**Comment by markusaksli-nc** [comment]
@darshankawar Don't actually have Linux with a touch screen. 

@jdebou What exact device is this on? Is it just a Linux laptop with touch or an embedded device?

**Comment by jdebou** [comment]
@markusaksli-nc I have two touchscreen laptops running ubuntu 20.04, and am experiencing this issue on both:
* [HP Elite Dragonfly](https://store.hp.com/FranceStore/Merch/Product.aspx?id=9WA26EA&opt=ABF&sel=NTB&lang=fr-FR&pup_e=1&pup_cid=33822&pup_id=9WA26EA%23ABF&gclid=Cj0KCQjwrsGCBhD1ARIsALILBYr2W9vmAvCa_1mSAcUJ6bY4gRJ-ITyZACpzfPG1eTUQZy-so_D1kY8aAkWUEALw_wcB&gclsrc=aw.ds)
* [Razer Blade pro 17](https://www.razer.com/fr-fr/gaming-laptops/Razer-Blade-Pro-17/RZ09-03295F63-R3F1)

**Comment by markusaksli-nc** [comment]
Labeling this for further investigation

**Comment by jdebou** [comment]
Small update: I am experiencing the same issue on a Microsoft Surface Book running Windows10. Should I open a separate issue to report on that?

**Comment by markusaksli-nc** [comment]
If it's the same issue then it might have the same root cause. We can just track it here.

**Comment by Hixie** [comment]
Does this work on other platforms (i.e. is it desktop-specific), or is it the same on Android and iOS?

**Comment by jdebou** [comment]
I do not experience this issue running the sample code on my android phone (pixel 4a).

**Comment by Hixie** [comment]
@chinmaygarde points out that it could be we have the wrong transform for the pointer events. Do things like tooltips work correctly? If you have a button in each corner that does something different, can you tap each button and that works correctly?

**Comment by Hixie** [comment]
@jdebou please see the comment above, thanks!

**Comment by jdebou** [comment]
Hi Hixie,

I tried buttons in each corner. They all work fine and so do tooltips.

<details>

```dart

import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}
class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {

    return Scaffold(
      body: Container(
        width: double.infinity,
        height: double.infinity,
        child: MyButtons(),
      ),
    );
  }
}

class MyButtons extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return MyButtonState();
  }
}

class MyButtonState extends State<MyButtons> {
  Color c;

  @override
  void initState() {
    super.initState();
    c = Colors.white70;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: c,
      child: Stack(
        children: [
          Positioned(
            top: 0,
            left: 0,
            child: _buildButton("blue", Colors.blue),
          ),
          Positioned(
            top: 0,
            right: 0,
            child: _buildButton("red", Colors.red),
          ),
          Positioned(
            bottom: 0,
            left: 0,
            child: _buildButton("green", Colors.green),
          ),
          Positioned(
            bottom: 0,
            right: 0,
            child: _buildButton("yellow", Colors.yellow),
          ),
        ],
      ),
    );
  }

  Widget _buildButton(String str, Color color) {
    return ElevatedButton(
      child: Tooltip(
        message: str,
        child: Container(height: 56, child: Text(str)),
      ),
      onPressed: () {
        setState(() {
          c = color;
        });
      },
    );
  }
}


```
</details>

Sorry about the delay, last week was wild. I should be more available now if you need me to run some more tests around this issue.

**Comment by Hixie** [comment]
This is really really weird, I don't understand how it's even possible. If we're transforming the points correctly, then the platform-specific code is working correctly, in which case, how can the platform-agnostic code (the gesture recognition code) even tell the difference?

**Comment by jdebou** [comment]
Could it be possible that the platform specific code provides bogus
position updates? I noticed that some position updates seem to be skipped
when using `GestureDetector.onPanUpdate`, making the sum of the updates
inconsistent with the actual position.

Le mar. 13 avr. 2021 à 22:54, Ian Hickson ***@***.***> a
écrit :

**Comment by jdebou** [comment]
Could it be possible that the platform specific code provides bogus
position updates? I noticed that some position updates seem to be skipped
when using `GestureDetector.onPanUpdate`, making the sum of the updates
inconsistent with the actual position.

**Comment by dnfield** [comment]
I'm able to reproduce this on a Windows laptop with a touchscreen.

I haven't proven, but strongly suspect, that we're just getting more sensitivity on the y-axis changes during a dominantly horizontal drag.

**Comment by dnfield** [comment]
Here's a small example of data packets that result in this:

```
flutter: PointerData(embedderId: 0, timeStamp: 441:11:31.592289, change: PointerChange.add, kind: PointerDeviceKind.mouse, signalKind: PointerSignalKind.none, device: 0, pointerIdentifier: 0, physicalX: 623.0, physicalY: 509.0, physicalDeltaX: 0.0, physicalDeltaY: 0.0, buttons: 0, synthesized: false, pressure: 0.0, pressureMin: 0.0, pressureMax: 0.0, distance: 0.0, distanceMax: 0.0, size: 0.0, radiusMajor: 0.0, radiusMinor: 0.0, radiusMin: 0.0, radiusMax: 0.0, orientation: 0.0, tilt: 0.0, platformData: 0, scrollDeltaX: 0.0, scrollDeltaY: 0.0)
flutter: PointerData(embedderId: 0, timeStamp: 441:11:31.592646, change: PointerChange.down, kind: PointerDeviceKind.mouse, signalKind: PointerSignalKind.none, device: 0, pointerIdentifier: 1, physicalX: 623.0, physicalY: 509.0, physicalDeltaX: 0.0, physicalDeltaY: 0.0, buttons: 1, synthesized: false, pressure: 0.0, pressureMin: 0.0, pressureMax: 0.0, distance: 0.0, distanceMax: 0.0, size: 0.0, radiusMajor: 0.0, radiusMinor: 0.0, radiusMin: 0.0, radiusMax: 0.0, orientation: 0.0, tilt: 0.0, platformData: 0, scrollDeltaX: 0.0, scrollDeltaY: 0.0)
flutter: PointerData(embedderId: 0, timeStamp: 441:11:31.592674, change: PointerChange.move, kind: PointerDeviceKind.mouse, signalKind: PointerSignalKind.none, device: 0, pointerIdentifier: 1, physicalX: 640.0, physicalY: 504.0, physicalDeltaX: 17.0, physicalDeltaY: -5.0, buttons: 1, synthesized: false, pressure: 0.0, pressureMin: 0.0, pressureMax: 0.0, distance: 0.0, distanceMax: 0.0, size: 0.0, radiusMajor: 0.0, radiusMinor: 0.0, radiusMin: 0.0, radiusMax: 0.0, orientation: 0.0, tilt: 0.0, platformData: 0, scrollDeltaX: 0.0, scrollDeltaY: 0.0)
flutter: vertical
```

It seems significant that we're getting the events as mouse rather than touch. @jonahwilliams says the logic for mouse is different.

**Comment by dnfield** [comment]
Yeah, if these do come through as touch it gets correctly detected as horizontal.

**Comment by dnfield** [comment]
The Windows and Linux embeddings both treat all pointer events as mouse events. See https://github.com/flutter/engine/blob/dade70a05f6a398169c80db20b0780b66f22d9f0/shell/platform/windows/flutter_windows_view.cc#L395 for Windows, and https://github.com/flutter/engine/blob/dade70a05f6a398169c80db20b0780b66f22d9f0/shell/platform/glfw/flutter_glfw.cc#L310 and https://github.com/flutter/engine/blob/dade70a05f6a398169c80db20b0780b66f22d9f0/shell/platform/linux/fl_engine.cc#L694 for Linux.

**Comment by dnfield** [comment]
Here's how to distinguish on Windows: https://docs.microsoft.com/en-us/windows/win32/tablet/system-events-and-mouse-messages?redirectedfrom=MSDN#distinguishing-pen-input-from-mouse-and-touch

**Comment by dnfield** [comment]
I'm having a little trouble setting up my Windows machine to build the engine - I'd need to install a bunch of software I don't have right now on it AFAICT.

@gspencergoog or @cbracken, is there someone that could take this over for that and/or Linux?

I created a quick patch here that I think should work, but needs to be tested/have tests written, and I'm not even sure if it compiles :) 
https://github.com/flutter/engine/compare/master..dnfield:win_touch

**Comment by stuartmorgan** [comment]
> The Windows and Linux embeddings both treat all pointer events as mouse events. 

Yes, touch was just never wired up. We also have open feature requests for multi-touch for both platforms.

**Comment by dnfield** [comment]
The Windows side of this will be fixed on master once the engine change rolls in. The GDK Linux side seems pretty fixable, but I'm not the right person to take on that work since I don't have access to a Linux device with a touch screen to test things.

**Comment by jdebou** [comment]
This is great news! Is there a way to help for the linux side?

**Comment by dnfield** [comment]
In the GDK embedding, I think we want to add https://developer.gnome.org/gdk3/stable/gdk3-Events.html#GDK-TOUCH-MASK:CAPS to https://github.com/flutter/engine/blob/dade70a05f6a398169c80db20b0780b66f22d9f0/shell/platform/linux/fl_view.cc#L193 and then wire through touch events.

Alternatively, it might be possible to somehow ask GDK if the mouse events it's simulating when we don't apply that mask are coming from a touch device rather than a mouse device. I can't really find anything about that in the docs though.

Basically, we just need to make sure that when we send the pointer data events to the framework, they have the right device kind in there so that the framework can figure out how to interpret the gestures.

**Comment by dnfield** [comment]
@cbracken might be interested in the Linux side

**Comment by gspencergoog** [comment]
cc @robert-ancell This may already be fixed, but maybe you have a touchscreen Linux device to test this on?

**Comment by robert-ancell** [comment]
> cc @robert-ancell This may already be fixed, but maybe you have a touchscreen Linux device to test this on?

I don't currently have any touch hardware, sorry.
